### ISSUE TITLE
Bug Report: Logical Error Rates with PAULI_CHANNEL_2

### ISSUE THREAD

**Issue by rsnegrin** [oscarhiggott/PyMatching#103]
### Description

I had an issue when using `PyMatching` decoder with circuits that include a `PAULI_CHANNEL_2` gate in the `Stim` framework. 

### Code
<details>
<summary>Expand to see the code</summary>

```python
import stim
import sinter
import os

p_r = 1e-6
p_cnot = 0.1
p2 = [
    p_cnot,  # p_ix
    0.0,
    0.0,
    0.0,
    0.0,
    0.0,
    0.0,
    0.0,
    0.0,
    0.0,
    0.0,
    0.0,
    0.0,
    0.0,
    0.0,
]

# Circuit with X_ERROR
circ1 = stim.Circuit()
circ1.append("R", range(3))
circ1.append("X_ERROR", range(3), p_r)
circ1.append("TICK")
circ1.append("CX", [0, 1])
circ1.append("X_ERROR", 1, p_cnot)
circ1.append("TICK")
circ1.append("CX", [2, 1])
circ1.append("X_ERROR", 1, p_cnot)
circ1.append("TICK")
circ1.append("M", range(3))

circ1.append("DETECTOR", [stim.target_rec(-3 + 1)], (1, 0))
circ1.append(
    "DETECTOR",
    [stim.target_rec(-3 + 0), stim.target_rec(-3 + 1), stim.target_rec(-3 + 2)],
    (0, 0),
)
circ1.append("OBSERVABLE_INCLUDE", [stim.target_rec(-1)], 0)

# Circuit with PAULI_CHANNEL_2
circ2 = stim.Circuit()
circ2.append("R", range(3))
circ2.append("X_ERROR", range(3), p_r)
circ1.append("TICK")
circ2.append("CX", [0, 1])
circ2.append("PAULI_CHANNEL_2", [0, 1], p2)
circ1.append("TICK")
circ2.append("CX", [2, 1])
circ2.append("PAULI_CHANNEL_2", [2, 1], p2)
circ1.append("TICK")
circ2.append("M", range(3))

circ2.append("DETECTOR", [stim.target_rec(-3 + 1)], (1, 0))
circ2.append(
    "DETECTOR",
    [stim.target_rec(-3 + 0), stim.target_rec(-3 + 1), stim.target_rec(-3 + 2)],
    (0, 0),
)
circ2.append("OBSERVABLE_INCLUDE", [stim.target_rec(-1)], 0)

tasks = [
    sinter.Task(circuit=circ, json_metadata={"i": i})
    for i, circ in enumerate([circ1, circ2])
]
results = sinter.collect(
    num_workers=os.cpu_count(),
    tasks=tasks,
    decoders=["pymatching"],
    max_shots=10_000_000,
    max_errors=None,
)
results.sort(key=lambda x: x.json_metadata["i"])

logical_error_rate1 = results[0].errors / results[0].shots
logical_error_rate2 = results[1].errors / results[1].shots
print(f"Logical error X_ERROR: {logical_error_rate1}")
print(f"Logical error PAULI_CHANNEL_2: {logical_error_rate2}")
```
</details>

Both circuits should be equivalent with the difference that one uses X_ERROR and the other PAULI_CHANNEL_2 to represent the same error.

### Output

Here es an example of the output: 

- **Logical error `X_ERROR`:** 1.5e-06
- **Logical error `PAULI_CHANNEL_2`:** 0.1799395

When using PAULI_CHANNEL_2 , I need to set `approximate_disjoint_errors=True` to be able to run the simulations. Sinter also sets  `decompose_errors=True` as default  [(details)](https://github.com/quantumlib/Stim/blob/main/glue/sample/src/sinter/_worker.py).

Here is the detector error model when using the PAULI_CHANNEL_2:

```python
circ2.detector_error_model(decompose_errors=True, approximate_disjoint_errors=True)

# Output:
stim.DetectorErrorModel('''
    error(1e-06) D0
    error(1e-06) D0 D1
    error(1e-06) D0 L0
    error(0.1) D1 L0 ^ D0 L0
    error(0.1) D1 ^ D0
    detector(1, 0) D0
    detector(0, 0) D1
''')

```

Other decoders like [Fusion Blossom](https://github.com/yuewuo/fusion-blossom) don’t have this problem and give similar results for both circuits. Also, the issue is not limited to X_ERROR, since I initially encountered this with Z_ERROR.

I'm not sure about the exact source of this problem, but it would be good to evaluate.
Thanks !

**Comment by oscarhiggott** [comment]
Hi @rsnegrin thanks for flagging this and apologies for the slow response. 

**TL;DR**: The discrepancy in logical error rate is mainly a consequence of a difference in how stim decomposes the detector error model in these two cases. If you choose not to decompose the detector error model (since it is already a graph), then the issue is resolved (both circuits have ~1e-6 logical error rates).

**In much more detail**:

Both circuits have the following undecomposed error model:
```
error(1e-06) D0
error(0.18000064) D0 D1
error(1e-06) D0 L0
detector(1, 0) D0
detector(0, 0) D1
```

Note that all three errors are already graphlike (either edges or boundary edges). If you ask stim to decompose circ1 (the one with X_ERROR), it does nothing, since all errors are edges. However, when you ask stim to decompose circ2 (the one with PAULI_CHANNEL_2) it splits the `error(0.18) D0 D1` edge into _three_ errors and decomposes only the second and third, each decomposed into a slightly different decomposition:

```
error(1e-06) D0
error(1e-06) D0 D1
error(1e-06) D0 L0
error(0.1) D1 L0 ^ D0 L0
error(0.1) D1 ^ D0
detector(1, 0) D0
detector(0, 0) D1
```

Specifically, the `error(1e-06) D0 D1` edge is left undecomposed and has low probability. The `error(0.18) D0 D1` edge is split into the decomposed error `error(0.1) D1 L0 ^ D0 L0` and the decomposed error `error(0.1) D1 ^ D0`. Importantly, both of these decomposed edges decompose into two boundary edge components, one of which is already in the DEM, and one of which is not. For `error(0.1) D1 L0 ^ D0 L0`, we see that the boundary edge component `error(p) D1 L0` is not actually in the DEM as an individual error, and for `error(0.1) D1 ^ D0` we see that `error(p) D1` is not already in the DEM.

Now if you give pymatching a decomposed DEM, it will treat each component as an edge, and all edges with the same detectors will be merged as if they were independent error mechanisms. Furthermore, if two edges have the same detectors but different observables, the first set of observables encountered will be kept (since the DEM must be distance 2 anyway, and parallel edges aren't supported). So when pymatching loads the decomposed DEM of circ2 if gets the following graph:
```
[
    (0, None, {'fault_ids': set(), 'weight': 1.52..., 'error_probability': 0.1800013...}), 
    (0, 1, {'fault_ids': set(), 'weight': 13.82..., 'error_probability': 1e-06}), 
    (1, None, {'fault_ids': {0}, 'weight': 1.52..., 'error_probability': 0.18})
]
```
equivalent to having loaded the following DEM:
```
error(0.1800013...) D0
error(1e-06) D0 D1
error(0.18) D1 L0
```
So in the common scenario (with probability around 0.18) where we observe the syndrome `D0 D1`, pymatching concludes it is more likely that we observed two high probability errors (`D0` and `D1 L0`) than one low probability `D0 D1` error, and predicts that `L0` was flipped, when in fact no observable was flipped. So this strange behavior is a result of stim's odd choice of decomposition for circ2 (decomposing an edge, and even doing so when one of the components is not already in the model), as well as the choice in pymatching to treat each _component_ as an independent edge.

*How to solve this problem*:

My suggestion is simply that you just don't try to decompose the DEM for this circuit. For example, you could change your sinter tasks to:
```
    tasks = [
        sinter.Task(
            circuit=circ,
            json_metadata={"i": i},
            detector_error_model=circ.detector_error_model(),
        )
        for i, circ in enumerate([circ1, circ2])
    ]
```


In terms of the pymatching behaviour, I think the current behaviour is already reasonable. I don't think the default behaviour should be to ignore decomposed errors, because you can already achieve that functionality by providing an undecomposed DEM. Also, in the case of surface code circuits with depolarizing noise, choosing to include the components in decompositions as edges can slightly improve accuracy. One other contributing factor to the issue here is that pymatching merges edges that have the same detectors but different observables - perhaps the assigned observables should be the more probable ones by default, however I would be hesitant to apply such a change if it had any impact on performance, since it only applies to circuits that are guaranteed to have distance at most 2 anyway.

The fact that stim decomposes an edge into two boundary edges when one of these does not exist seems like a potential problem with the decomposition, not the decoder. I'm not sure if Craig would be open to changing the decomposition rules in stim, although I suspect this design choice is there for a reason.

I'm closing this issue for now as I don't think it motivates a change to pymatching, but please let me know if you have any questions.
